### PR TITLE
Preserve url hash for saml based authentication

### LINKED ISSUE #831: [BUG] SAML login nextUrl drops URL hash
**Describe the bug**
[OpenSearch Dashboards encodes useful state in the `hash` component of the URL](https://www.elastic.co/guide/en/kibana/master/kibana-navigation.html#state-sync), for example:

```
https://<domain>/app/discover#/?_a=(columns:!(level,loggerCategory,messageText,shardName,tags.subDir,workspaceId))
```

Currently, the SAML component's `nextUrl` builder includes `url.path` but omits `url.hash`, dropping this state.  This creates a problem when sharing links to a particular view without first creating a shortlink, in that the state of that view is lost.  For the example above, the redirect would be:

```
https://<domain>/auth/saml/login?nextUrl=%2Fapp%2Fdiscover
```

and after authentication, the user would land at:

```
https://<domain>/app/discover
```

With no preselected columns.

**Expected behavior**
I believe the user expect to be redirected back to the exact URL, including hash, as they were at before the authentication redirect.  I believe it would be preferable to include `url.hash`, so the `nextUrl` would become:

```
https://<domain>/auth/saml/login?nextUrl=%2Fapp%2Fdiscover%23%2F%3F_a%3D%28columns%3A%21%28level%2CloggerCategory%2CmessageText%2CshardName%2Ctags.subDir%2CworkspaceId%29%29
```

**Additional context**
 * https://github.com/opensearch-project/security-dashboards-plugin/blob/afd64971de308b49622640cbc194d717cc802a17/server/auth/types/saml/saml_auth.ts#L57
 * https://nodejs.org/dist/latest-v8.x/docs/api/url.html#url_url_strings_and_url_objects
 * https://www.elastic.co/guide/en/kibana/master/kibana-navigation.html#state-sync

### PR DESCRIPTION
**Pull request by devardee**
### Description
Preserves the url hash after saml handshake.

### Category
Enhancement

### Why these changes are required?
For better user experience. refer to #543 and #831. 

### What is the old behavior before changes and new behavior after changes?
 refer to #543 and #831

### Issues Resolved
#831 #543 

### Testing
manual tested.

### Check List
- [Yes ] New functionality includes testing
- [No ] New functionality has been documented
- [ Yes] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by peternied — CHANGES_REQUESTED** [review]
Thank you for the contribution, this will greatly improve the post sign-in experience with SAML auth.

There are a couple of gaps before this change can be merged.  Functional test cases for the new routes and the end to end workflow must be added.

All commits are required to be signed off per the DCO [more details](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Let us know if you have questions about how to accomplish these outstanding items,

**Comment by cliu123** [comment]
@devardee Have you got chance to follow up on this PR at all? We would love to have another review when the comments are resolved. Thanks!

**Comment by cliu123** [comment]
Closing the PR due to inactivity for now. Please feel free to re-open if needed.

**Pull request closed without merging**
